pkg/reporter/reporter.go

```diff
@@ -423,3 +436,41 @@ func getUniqueViolationURLs(violations []report.Violation) map[string]string {
 
 	return urls
 }
+
+// Publish prints a JUnit XML report to the configured output.
+func (tr JUnitReporter) Publish(_ context.Context, r report.Report) error {
+	testSuites := junit.Testsuites{
+		Name: "regal",
+	}
+
+	// group by file & sort by file
+	files := make([]string, 0)
+	violationsPerFile := map[string][]report.Violation{}
+
+	for _, violation := range r.Violations {
+		files = append(files, violation.Location.File)
+		violationsPerFile[violation.Location.File] = append(violationsPerFile[violation.Location.File], violation)
+	}
+
+	sort.Strings(files)
+
+	for _, file := range files {
+		testsuite := junit.Testsuite{
+			Name: file,
+		}
+
+		for _, violation := range violationsPerFile[file] {
+			testsuite.AddTestcase(junit.Testcase{
+				Name:      violation.Title,
+				Classname: violation.Location.String(),
+				Failure: &junit.Result{
+					Message: fmt.Sprintf("%s. To learn more, see: %s", violation.Description, getDocumentationURL(violation)),
```

Comment:
Most notable omission here would be the category... and understandably so given that it's a test result format after all. If there isn't some clever meta attribute we can use for that, perhaps we could format the `Name` like `$category/$title: $message`, or perhaps have the `Message` include it?

Knowing the category helps with prioritization of fixing the issues reported, so it'd be good to have it in there _somewhere_, but again, if there's some other place to put it, I'm open to suggestions.

Reply:
oh yeah good idea! What would `$message` be though? Maybe just the description of a violation and have the `Message` contain both description and link to the docs?

I think we can add any kind of additional metadata to the `Data` field of a `junit.Result`. It's currently not used in the code but we could write all the violation details in there. I'll try it out on our instance and post screenshots to further the discussion

Reply:
This shows the test report widget inside a merge request:

![2024-07-22T10:15:42,541965099+02:00](https://github.com/user-attachments/assets/a7805a02-1903-41af-beab-d2fd875d0732)

that's the popup you get when clicking on `View details`:

![2024-07-22T10:15:59,187202303+02:00](https://github.com/user-attachments/assets/7f8cfc26-8baf-4656-a80d-f33009d8d238)

and this is what the `Full report` looks like

![2024-07-22T10:16:15,244480379+02:00](https://github.com/user-attachments/assets/ee039643-7e18-4b7e-93a8-7951f804c8f8)

when you click on `View details` in the full report, you get the same popup as in the merge request

Reply:
Looks great! Where does the category come from there? You've updated your code to show that? I don't see it in the PR, but that looks great to me, so please do include it!

Reply:
er yeah I pushed it to another branch so not to pollute this PR. I also just realized that GitLab no longer shows the failure message which included the link to the documentation. So the latest version now includes the documentation link and its output is aligned with the pretty printer. It looks like this:

![2024-07-22T12:50:15,296173773+02:00](https://github.com/user-attachments/assets/92c16317-7b9d-48fe-9796-58ffa609a304)

It's kinda sad that GitLab formats this output as a code block so links are not clickable but it think there is no way around that

Reply:
Yeah, but not much to do about it I suppose. GitHub doesn't make it a link either, even if it's not in a codeblock. In that context we can at least display real markdown links in the summary though.

Reply:
Looks great though!